keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+	// that is used when creating the resources.
+	// +optional
+	APIGroup string
+
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+
+	// FieldPath is a dot separarated JSON path that defines where in the
+	// parameter object the numeric parameters are embedded.
+	FieldPath string
+
+	// Shareable is copied into the AllocationResult when a claim
+	// gets allocated.
+	Shareable bool
+}
+```
+
+### NodeResourceCapacity
+
+For each node, a NodeResourceCapacity objects gets created:
+
+```
+// NodeResourceCapacity gets published by kubelet. Its name
+// matches the name of the node and the node is the owner.
+//
+// Capacity may get added, but should not get removed because
+// it would make scheduling decisions based on the old capacity
+// invalid.
+type NodeResourceCapacity struct {
+	metav1.TypeMeta
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Resources contains information about the capacity
+	// reported by each DRA driver.
+	//
+	// +optional
+	Resources []DriverResources
+}
+
+type DriverResources struct {
+	// DriverName identifies the DRA which provided the capacity information.
+	DriverName string
+
+	// ResourceInstances describes all discrete resource instances that are
+	// managed by the driver. Each entry must be an object of one of the
+	// supported numeric resource capacity types with kind, version and uid
+	// set. The uid only needs to be unique inside the node.
+	//
+	// +optional
+	ResourceInstances []runtime.RawExtension
+}
+```
+
+If a driver needs to remove capacity, then there is a risk that a claim gets
+allocated using that capacity while the kubelet is updating the
+NodeResourceCapacity object. The implementations of numeric models must handle
+scenarios where more resources are allocated than available. The DRA driver
+kubelet plugin must double-check that the allocated resources are still
+available when NodePrepareResource is called. If not, the pod cannot cannot
+start until the resource comes back.
+
+### Builtin controllers
+
+`k8s.io/dynamic-resource-allocation/builtincontroller` defines Go interfaces
+that need to be implemented by a numeric model to handle tracking of resource
+usage and allocation/deallocation. The package also has a global registry of
+builtin controllers.
+
+##### Initialization
+
+All registered builtin controllers get activated when kube-scheduler or CA
+start. At this point they start building an internal model of which resources
+are available (based on those entries in NodeResourceCapacity objects that they
+are responsible for) and how those resources are already used (based on the
+status of allocated claims).
+
+##### Allocation
+
+When the dynamic resource plugin in kube-scheduler encounters a claim which,
+according to the class, uses numeric parameters, it looks up which builtin
+controller handles the claim parameters. That controller then is asked to
```

Comment:
I must be missing something still.  How does this lookup happen? Meaning what are the criteria for lookup?

Reply:
I need to be more explicit here. It's still evolving a bit, though, with the actual API changes in the ResourceClass still not finalized (field path -> Go template -> CEL).

Right now, each builtin controller is invoked and then checks whether it understands the claim parameter type after the conversion to the numeric model. The one which does is used. There will be at most one controller per type.

There might be none. This can happen when the numeric type expected by the DRA driver is not enabled or not available in the scheduler. Then this claim cannot be allocated and the scheduler will emit an event with that information.